storage/src/mmr/hasher.rs

```diff
@@ -1,40 +1,76 @@
 //! Decorator for a cryptographic hasher that implements the MMR-specific hashing logic.
 
+use std::future::Future;
+
+use crate::mmr::{Error, Storage};
 use commonware_cryptography::Hasher as CHasher;
 
+/// A trait required by various MMR methods for computing hash digests.
+pub trait Hasher<H: CHasher>: Send + Sync {
+    /// Computes the hash for a leaf given its position and the element it represents.
+    fn leaf_hash(
+        &mut self,
+        pos: u64,
+        element: &[u8],
+    ) -> impl Future<Output = Result<H::Digest, Error>> + Send;
+
+    /// Computes the hash for a node given its position and the hashes of its children.
+    fn node_hash(&mut self, pos: u64, left_hash: &H::Digest, right_hash: &H::Digest) -> H::Digest;
+
+    /// Computes the root hash for an MMR given its size and an iterator over the hashes of its
+    /// peaks. The iterator should yield the peak hashes in decreasing order of their height.
+    fn root_hash<'b>(
+        &mut self,
+        pos: u64,
+        peak_hashes: impl Iterator<Item = &'b H::Digest>,
+    ) -> H::Digest;
+
+    fn hash(&mut self, data: &[u8]) -> H::Digest;
+}
+
 /// Hasher decorator the MMR uses for computing leaf, node and root hashes.
-pub(crate) struct Hasher<'a, H: CHasher> {
-    hasher: &'a mut H,
+pub struct Basic<H: CHasher> {
+    hasher: H,
 }
 
-impl<'a, H: CHasher> Hasher<'a, H> {
-    pub(crate) fn new(hasher: &'a mut H) -> Self {
+impl<H: CHasher> Basic<H> {
+    /// Creates a new `Basic` hasher.
+    pub fn new(hasher: H) -> Self {
         Self { hasher }
     }
 
-    /// Computes the hash for a leaf given its position and the element it represents.
-    pub(crate) fn leaf_hash(&mut self, pos: u64, element: &[u8]) -> H::Digest {
+    pub(crate) fn update_with_pos(&mut self, pos: u64) {
+        self.hasher.update(&pos.to_be_bytes());
+    }
+
+    pub(crate) fn update_with_hash(&mut self, hash: &H::Digest) {
+        self.hasher.update(hash.as_ref());
+    }
+
+    pub(crate) fn update_with_element(&mut self, element: &[u8]) {
+        self.hasher.update(element);
+    }
+
+    pub(crate) fn finalize_reset(&mut self) -> H::Digest {
```

Comment:
consistency: just called this `finalize()`

Reply:
done